File: Core kernel razonamiento repo para Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/go/github_token_provider_test.go
package copilot

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/github/copilot-sdk/go/internal/jsonrpc2"
	"github.com/github/copilot-sdk/go/rpc"
)

func TestGitHubTokenProviderConfigValidation(t *testing.T) {
	provider := func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenCancelled(), nil
	}

	if _, err := NewClient(nil).CreateSession(t.Context(), &SessionConfig{
		GitHubToken:         "static",
		GitHubTokenProvider: provider,
	}); err == nil || !strings.Contains(err.Error(), "cannot be used together") {
		t.Fatalf("CreateSession error = %v", err)
	}
	if _, err := NewClient(nil).ResumeSession(t.Context(), "session", &ResumeSessionConfig{
		GitHubToken:         "static",
		GitHubTokenProvider: provider,
	}); err == nil || !strings.Contains(err.Error(), "cannot be used together") {
		t.Fatalf("ResumeSession error = %v", err)
	}
}

func TestGitHubTokenProviderCreateRequestAndCallback(t *testing.T) {
	rpcClient, server, _ := newRuntimeShutdownRpcPair(t)
	t.Cleanup(server.Stop)
	client := &Client{
		client:   rpcClient,
		RPC:      rpc.NewServerRPC(rpcClient),
		sessions: make(map[string]*Session),
	}
	client.setupNotificationHandler()

	var createParams json.RawMessage
	server.SetRequestHandler("session.create", func(params json.RawMessage) (json.RawMessage, *jsonrpc2.Error) {
		createParams = append(json.RawMessage(nil), params...)
		sessionID := sessionIDFromParams(t, params)
		return []byte(`{"sessionId":"` + sessionID + `","workspacePath":"/workspace"}`), nil
	})
	server.SetRequestHandler("session.destroy", func(json.RawMessage) (json.RawMessage, *jsonrpc2.Error) {
		return []byte(`{}`), nil
	})

	var gotArgs GitHubTokenProviderArgs
	session, err := client.CreateSession(t.Context(), &SessionConfig{
		GitHubTokenProvider: func(args GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
			gotArgs = args
			return GitHubTokenResult(&GitHubToken{
				AccessToken: "secret-token",
				TokenType:   String("bearer"),
				ExpiresIn:   8 * 60 * 60,
			}), nil
		},
	})
	if err != nil {
		t.Fatalf("CreateSession failed: %v", err)
	}

	var wire struct {
		RegistrationID string `json:"gitHubTokenProviderRegistrationId"`
		GitHubToken    string `json:"gitHubToken"`
	}
	if err := json.Unmarshal(createParams, &wire); err != nil {
		t.Fatal(err)
	}
	if wire.RegistrationID == "" {
		t.Fatal("gitHubTokenProviderRegistrationId was not serialized")
	}
	if wire.GitHubToken != "" {
		t.Fatal("static gitHubToken should not be serialized")
	}

	sessionID := session.SessionID
	raw, rpcErr := server.Request(t.Context(), "gitHubToken.getToken", &rpc.GitHubTokenAcquireRequest{
		RegistrationID: wire.RegistrationID,
		Host:           "github.example.com",
		SessionID:      &sessionID,
		Reason:         rpc.GitHubTokenAcquireReasonRefresh,
	})
	if rpcErr != nil {
		t.Fatalf("getToken failed: %v", rpcErr)
	}
	var tokenResult struct {
		Kind        string `json:"kind"`
		AccessToken string `json:"accessToken"`
		ExpiresIn   int64  `json:"expiresIn"`
	}
	if err := json.Unmarshal(raw, &tokenResult); err != nil {
		t.Fatal(err)
	}
	if tokenResult.Kind != "token" || tokenResult.AccessToken != "secret-token" || tokenResult.ExpiresIn != 8*60*60 {
		t.Fatalf("unexpected token result: %+v", tokenResult)
	}
	if gotArgs.Host != "github.example.com" || gotArgs.SessionID == nil ||
		*gotArgs.SessionID != sessionID || gotArgs.Reason != GitHubTokenRequestReasonRefresh {
		t.Fatalf("unexpected callback args: %+v", gotArgs)
	}

	if err := session.Disconnect(); err != nil {
		t.Fatal(err)
	}
	if len(client.gitHubTokenProviders) != 0 {
		t.Fatal("provider registration was not removed on disconnect")
	}
	if _, rpcErr := server.Request(t.Context(), "gitHubToken.getToken", &rpc.GitHubTokenAcquireRequest{
		RegistrationID: wire.RegistrationID,
		Host:           "github.com",
		Reason:         rpc.GitHubTokenAcquireReasonInitial,
	}); rpcErr == nil {
		t.Fatal("unknown registration ID should return a handler error")
	}

	var resumeParams json.RawMessage
	server.SetRequestHandler("session.resume", func(params json.RawMessage) (json.RawMessage, *jsonrpc2.Error) {
		resumeParams = append(json.RawMessage(nil), params...)
		return []byte(`{"sessionId":"resumed-session","workspacePath":"/workspace"}`), nil
	})
	resumed, err := client.ResumeSession(t.Context(), "resumed-session", &ResumeSessionConfig{
		GitHubTokenProvider: func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
			return GitHubTokenCancelled(), nil
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(resumeParams, &wire); err != nil {
		t.Fatal(err)
	}
	if wire.RegistrationID == "" {
		t.Fatal("resume did not serialize gitHubTokenProviderRegistrationId")
	}
	if err := resumed.Disconnect(); err != nil {
		t.Fatal(err)
	}
}

func TestGitHubTokenProviderResultsErrorsAndRollback(t *testing.T) {
	client := &Client{}
	adapter := &gitHubTokenAdapter{client: client}

	cancelID := client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenCancelled(), nil
	})
	result, err := adapter.GetToken(&rpc.GitHubTokenAcquireRequest{RegistrationID: cancelID})
	if err != nil {
		t.Fatal(err)
	}
	if _, ok := result.(*rpc.GitHubTokenAcquireResultCancelled); !ok {
		t.Fatalf("result type = %T", result)
	}

	sentinel := errors.New("provider failed")
	errorID := client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return nil, sentinel
	})
	if _, err := adapter.GetToken(&rpc.GitHubTokenAcquireRequest{RegistrationID: errorID}); !errors.Is(err, sentinel) {
		t.Fatalf("provider error = %v", err)
	}

	rpcClient, server, _ := newRuntimeShutdownRpcPair(t)
	t.Cleanup(server.Stop)
	rollbackClient := &Client{
		client:   rpcClient,
		RPC:      rpc.NewServerRPC(rpcClient),
		sessions: make(map[string]*Session),
	}
	server.SetRequestHandler("session.create", func(json.RawMessage) (json.RawMessage, *jsonrpc2.Error) {
		return nil, &jsonrpc2.Error{Code: -32000, Message: "create failed"}
	})
	if _, err := rollbackClient.CreateSession(t.Context(), &SessionConfig{
		GitHubTokenProvider: func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
			return GitHubTokenCancelled(), nil
		},
	}); err == nil {
		t.Fatal("expected create failure")
	}
	if len(rollbackClient.gitHubTokenProviders) != 0 {
		t.Fatal("provider registration was not rolled back")
	}
}

func TestGitHubTokenStringRedactsAccessToken(t *testing.T) {
	token := GitHubToken{AccessToken: "secret-token", ExpiresIn: 28_800}

	if got := fmt.Sprintf("%v %#v", token, token); strings.Contains(got, token.AccessToken) {
		t.Fatalf("GitHubToken formatting exposed the access token: %s", got)
	}
}

func TestGitHubTokenProviderCleanupOnDisconnectError(t *testing.T) {
	rpcClient, server, _ := newRuntimeShutdownRpcPair(t)
	t.Cleanup(server.Stop)
	server.SetRequestHandler("session.destroy", func(json.RawMessage) (json.RawMessage, *jsonrpc2.Error) {
		return nil, &jsonrpc2.Error{Code: -32000, Message: "destroy failed"}
	})
	client := &Client{}
	registrationID := client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenCancelled(), nil
	})
	session := newSession("cleanup-session", rpcClient, "", false)
	session.setGitHubTokenProviderRegistrationRelease(func() {
		client.unregisterGitHubTokenProvider(registrationID)
	})

	if err := session.Disconnect(); err == nil || !strings.Contains(err.Error(), "destroy failed") {
		t.Fatalf("Disconnect error = %v", err)
	}
	if len(client.gitHubTokenProviders) != 0 {
		t.Fatal("provider registration was not removed after disconnect failed")
	}
}

func TestGitHubTokenProviderReleaseBeforeOwnershipTransfer(t *testing.T) {
	client := &Client{}
	registrationID := client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenCancelled(), nil
	})
	session := &Session{}

	session.releaseGitHubTokenProviderRegistration()
	session.setGitHubTokenProviderRegistrationRelease(func() {
		client.unregisterGitHubTokenProvider(registrationID)
	})

	if len(client.gitHubTokenProviders) != 0 {
		t.Fatal("provider registration was not removed after a pending session had already been retired")
	}
}

func TestGitHubTokenProviderCleanupOnDelete(t *testing.T) {
	rpcClient, server, _ := newRuntimeShutdownRpcPair(t)
	t.Cleanup(server.Stop)
	server.SetRequestHandler("session.delete", func(json.RawMessage) (json.RawMessage, *jsonrpc2.Error) {
		return []byte(`{"success":true}`), nil
	})
	client := &Client{
		client:   rpcClient,
		sessions: make(map[string]*Session),
		state:    stateConnected,
	}
	registrationID := client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenCancelled(), nil
	})
	session := newSession("delete-session", rpcClient, "", false)
	session.setGitHubTokenProviderRegistrationRelease(func() {
		client.unregisterGitHubTokenProvider(registrationID)
	})
	client.sessions[session.SessionID] = session

	if err := client.DeleteSession(t.Context(), session.SessionID); err != nil {
		t.Fatal(err)
	}
	if len(client.gitHubTokenProviders) != 0 {
		t.Fatal("provider registration was not removed after session deletion")
	}
}

func TestSessionOperationsSerializeBySessionID(t *testing.T) {
	client := &Client{}
	unlockFirst := client.lockSessionOperation("same-session")
	sameSessionAcquired := make(chan struct{})
	go func() {
		unlock := client.lockSessionOperation("same-session")
		close(sameSessionAcquired)
		unlock()
	}()

	select {
	case <-sameSessionAcquired:
		t.Fatal("same-session operation was not serialized")
	case <-time.After(25 * time.Millisecond):
	}

	otherSessionAcquired := make(chan struct{})
	go func() {
		unlock := client.lockSessionOperation("other-session")
		close(otherSessionAcquired)
		unlock()
	}()
	select {
	case <-otherSessionAcquired:
	case <-time.After(time.Second):
		t.Fatal("different-session operation was unnecessarily blocked")
	}

	unlockFirst()
	select {
	case <-sameSessionAcquired:
	case <-time.After(time.Second):
		t.Fatal("same-session operation did not proceed after release")
	}
}

func TestGitHubTokenProvidersClearedOnConnectionClose(t *testing.T) {
	client := &Client{state: stateConnected}
	client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenCancelled(), nil
	})

	client.handleConnectionClose()

	if len(client.gitHubTokenProviders) != 0 {
		t.Fatal("provider registrations were not cleared after connection closure")
	}
}

func TestGitHubTokenProviderConcurrentRegistrationsAreIsolated(t *testing.T) {
	client := &Client{}
	adapter := &gitHubTokenAdapter{client: client}
	idA := client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenResult(&GitHubToken{AccessToken: "a", ExpiresIn: 1}), nil
	})
	idB := client.registerGitHubTokenProvider(func(GitHubTokenProviderArgs) (*GitHubTokenProviderResult, error) {
		return GitHubTokenResult(&GitHubToken{AccessToken: "b", ExpiresIn: 1}), nil
	})

	var wg sync.WaitGroup
	for id, want := range map[string]string{idA: "a", idB: "b"} {
		wg.Add(1)
		go func() {
			defer wg.Done()
			result, err := adapter.GetToken(&rpc.GitHubTokenAcquireRequest{RegistrationID: id})
			if err != nil {
				t.Error(err)
				return
			}
			if got := result.(*rpc.GitHubTokenAcquireResultToken).AccessToken; got != want {
				t.Errorf("token = %q, want %q", got, want)
			}
		}()
	}
	wg.Wait()
}

File: Core kernel razonamiento repo para Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/nodejs/test/github-token-provider.test.ts
import { describe, expect, it, vi } from "vitest";
import { CopilotClient, RuntimeConnection, type GitHubTokenProvider } from "../src/index.js";

function createMockClient(
    request: (method: string, params: Record<string, unknown>) => Promise<unknown>
): CopilotClient {
    const client = new CopilotClient({
        connection: RuntimeConnection.forUri("localhost:1234"),
    });
    (client as unknown as { connection: unknown }).connection = {
        sendRequest: request,
        dispose: vi.fn(),
    };
    return client;
}

function getTokenHandler(client: CopilotClient) {
    return (
        client as unknown as {
            clientGlobalHandlers: {
                gitHubToken: {
                    getToken(params: {
                        registrationId: string;
                        host: string;
                        sessionId?: string;
                        reason: "initial" | "refresh";
                    }): Promise<unknown>;
                };
            };
        }
    ).clientGlobalHandlers.gitHubToken.getToken;
}

describe("session GitHub token providers", () => {
    it("rejects a static token and provider together", async () => {
        const client = new CopilotClient({
            connection: RuntimeConnection.forUri("localhost:1234"),
        });

        await expect(
            client.createSession({
                gitHubToken: "static",
                gitHubTokenProvider: async () => ({
                    kind: "token",
                    accessToken: "dynamic",
                    expiresIn: 28_800,
                }),
            })
        ).rejects.toThrow("gitHubToken and gitHubTokenProvider are mutually exclusive");
    });

    it("serializes only the opaque registration and maps token and cancellation results", async () => {
        let createPayload: Record<string, unknown> | undefined;
        const request = vi.fn(async (method: string, params: Record<string, unknown>) => {
            if (method === "session.create") {
                createPayload = params;
                return { sessionId: params.sessionId };
            }
            throw new Error(`Unexpected method: ${method}`);
        });
        const observed: unknown[] = [];
        const provider: GitHubTokenProvider = vi
            .fn()
            .mockImplementationOnce(async (args) => {
                observed.push(args);
                return {
                    kind: "token",
                    accessToken: "secret-token",
                    tokenType: "Bearer",
                    expiresIn: 28_800,
                };
            })
            .mockImplementationOnce(async (args) => {
                observed.push(args);
                return { kind: "cancelled" };
            });
        const client = createMockClient(request);
        const session = await client.createSession({
            sessionId: "session-one",
            gitHubTokenProvider: provider,
        });

        const registrationId = createPayload?.gitHubTokenProviderRegistrationId;
        expect(registrationId).toMatch(
            /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/
        );
        expect(createPayload).not.toHaveProperty("gitHubTokenProvider");
        expect(createPayload?.gitHubToken).toBeUndefined();

        const handler = getTokenHandler(client);
        await expect(
            handler({
                registrationId: registrationId as string,
                host: "github.example.com",
                reason: "initial",
            })
        ).resolves.toEqual({
            kind: "token",
            accessToken: "secret-token",
            tokenType: "Bearer",
            expiresIn: 28_800,
        });
        await expect(
            handler({
                registrationId: registrationId as string,
                host: "github.example.com",
                sessionId: session.sessionId,
                reason: "refresh",
            })
        ).resolves.toEqual({ kind: "cancelled" });
        expect(observed).toEqual([
            {
                host: "github.example.com",
                sessionId: "session-one",
                reason: "initial",
            },
            {
                host: "github.example.com",
                sessionId: "session-one",
                reason: "refresh",
            },
        ]);
    });

    it("preserves callback errors and rejects unknown registrations", async () => {
        const client = createMockClient(async (_method, params) => ({
            sessionId: params.sessionId,
        }));
        const failure = new Error("credential broker failed");
        await client.createSession({
            sessionId: "error-session",
            gitHubTokenProvider: () => {
                throw failure;
            },
        });
        const registrationId = [
            ...(
                client as unknown as {
                    githubTokenProviders: Map<string, unknown>;
                }
            ).githubTokenProviders.keys(),
        ][0];
        const handler = getTokenHandler(client);

        await expect(
            handler({
                registrationId,
                host: "github.com",
                reason: "initial",
            })
        ).rejects.toBe(failure);
        await expect(
            handler({
                registrationId: "unknown",
                host: "github.com",
                reason: "refresh",
            })
        ).rejects.toThrow("No GitHub token provider registered");
    });

    it("rolls back failed creation and cleans up on session and client close", async () => {
        const failingClient = createMockClient(async () => {
            throw new Error("create failed");
        });
        await expect(
            failingClient.createSession({
                gitHubTokenProvider: async () => ({ kind: "cancelled" }),
            })
        ).rejects.toThrow("create failed");
        expect(
            (
                failingClient as unknown as {
                    githubTokenProviders: Map<string, unknown>;
                }
            ).githubTokenProviders
        ).toHaveLength(0);

        const client = createMockClient(async (method, params) => {
            if (method === "session.create") return { sessionId: params.sessionId };
            if (method === "session.destroy") return {};
            if (method === "session.delete") return { success: true };
            throw new Error(`Unexpected method: ${method}`);
        });
        const first = await client.createSession({
            sessionId: "first",
            gitHubTokenProvider: async () => ({ kind: "cancelled" }),
        });
        await client.createSession({
            sessionId: "second",
            gitHubTokenProvider: async () => ({ kind: "cancelled" }),
        });
        const registrations = (
            client as unknown as {
                githubTokenProviders: Map<string, unknown>;
            }
        ).githubTokenProviders;
        expect(registrations).toHaveLength(2);

        await first.disconnect();
        expect(registrations).toHaveLength(1);
        await client.deleteSession("second");
        expect(registrations).toHaveLength(0);
        await client.forceStop();
        expect(registrations).toHaveLength(0);
    });

    it("rotates a resumed session only after resume succeeds", async () => {
        const payloads: Record<string, unknown>[] = [];
        const client = createMockClient(async (method, params) => {
            payloads.push(params);
            if (method === "session.create" || method === "session.resume") {
                return { sessionId: params.sessionId };
            }
            throw new Error(`Unexpected method: ${method}`);
        });
        const firstProvider = vi.fn(async () => ({ kind: "cancelled" as const }));
        const secondProvider = vi.fn(async () => ({ kind: "cancelled" as const }));
        await client.createSession({
            sessionId: "resumed",
            gitHubTokenProvider: firstProvider,
        });
        const firstRegistration = payloads[0].gitHubTokenProviderRegistrationId as string;

        await client.resumeSession("resumed", {
            gitHubTokenProvider: secondProvider,
        });
        const secondRegistration = payloads[1].gitHubTokenProviderRegistrationId as string;
        const handler = getTokenHandler(client);

        await expect(
            handler({
                registrationId: firstRegistration,
                host: "github.com",
                reason: "refresh",
            })
        ).rejects.toThrow("No GitHub token provider registered");
        await expect(
            handler({
                registrationId: secondRegistration,
                host: "github.com",
                reason: "refresh",
            })
        ).resolves.toEqual({ kind: "cancelled" });
        expect(secondProvider).toHaveBeenCalledOnce();
        expect(firstProvider).not.toHaveBeenCalled();
    });

    it("does not retire a concurrent pending registration", async () => {
        const resumeResolvers: Array<(value: { sessionId: string }) => void> = [];
        const payloads: Record<string, unknown>[] = [];
        const client = createMockClient(async (method, params) => {
            payloads.push(params);
            if (method === "session.create") {
                return { sessionId: params.sessionId };
            }
            if (method === "session.resume") {
                return await new Promise<{ sessionId: string }>((resolve) => {
                    resumeResolvers.push(resolve);
                });
            }
            throw new Error(`Unexpected method: ${method}`);
        });
        await client.createSession({
            sessionId: "concurrent",
            gitHubTokenProvider: async () => ({ kind: "cancelled" }),
        });

        const firstResume = client.resumeSession("concurrent", {
            gitHubTokenProvider: async () => ({ kind: "cancelled" }),
        });
        const secondResume = client.resumeSession("concurrent", {
            gitHubTokenProvider: async () => ({ kind: "cancelled" }),
        });
        await vi.waitFor(() => expect(resumeResolvers).toHaveLength(2));

        resumeResolvers[0]({ sessionId: "concurrent" });
        await firstResume;
        const registrations = (
            client as unknown as {
                githubTokenProviders: Map<string, unknown>;
            }
        ).githubTokenProviders;
        expect(registrations).toHaveLength(2);

        resumeResolvers[1]({ sessionId: "concurrent" });
        await secondResume;
        expect(registrations).toHaveLength(1);
        expect(registrations.has(payloads[2].gitHubTokenProviderRegistrationId as string)).toBe(
            true
        );
    });
});
